### PR TITLE
Make versions their own concrete objects, and allow time travel.

### PR DESCRIPTION
**Pull request by ndmckinley**
### Description
This change implements #806.

It is huge.  Here's the reading order I suggest:

1. conf/evolutions - see the changes to the database.
2. models/Version.java - see the new Version object.
3. models/* - see how adding versions changes the other resources.
4. repository/VersionRepository.java
5. controllers/admin/AdminVersionController.java (and associated view)
6. other repositories
7. other services
8. all the test files.

Here's a short summary:
We create a "version" object, and each program and question can be in one or more versions.  We present those versions to the admin by ID and date.  The admin can select any version to go live.

We modify the program and question by removing lifecycle stage (which accounts for most of the change by volume) and version (much of the rest).  Those concepts are held in the Version object now.  Anyplace we would have queried for lifecycle stage is now a two-step query - selecting a version then finding whether it contains the program or question.  This is mildly less efficient but it's much clearer in code than the alternative (which involves some strange-looking ebeans magic we don't use elsewhere).

We add tests for the version repository and adapt all the existing tests for the program and question services and repositories.

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
When merged, this will trigger completion review of #759.

### CODE REVIEW

**Comment by bion** [comment]
On the plus side, more code is removed than added!

File: universal-application-tool-0.0.1/app/models/Version.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.List;
+import java.util.Optional;
+import javax.persistence.Entity;
+import javax.persistence.ManyToMany;
+import javax.persistence.Table;
+import play.data.validation.Constraints;
+
+@Entity
+@Table(name = "versions")
+public class Version extends BaseModel {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/models/Version.java:15]: suggestion (blocking): add javadoc for this class

File: universal-application-tool-0.0.1/app/models/Version.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  public Instant getSubmitTime() {
+    return this.submitTime;
+  }
+
+  public Optional<Program> getProgramByName(String name) {
+    return getPrograms().stream()
+        .filter(p -> p.getProgramDefinition().adminName().equals(name))
+        .findAny();
```
Comment (bion) [inline universal-application-tool-0.0.1/app/models/Version.java:57]: Question (blocking): should this be `findOne`? What guarantees are there about the program this finds? Please document this and `getQuestionByName`

File: universal-application-tool-0.0.1/app/controllers/admin/QuestionController.java [hunk 3]
```diff
@@ -191,14 +184,7 @@ public Result update(Request request, Long id, String questionType) {
...
           service.getReadOnlyQuestionService().toCompletableFuture().join();
-      questionDefinition =
-          getBuilderWithQuestionPath(roService, questionForm)
-              .setId(id)
-              // Version is needed for building a question definition.
-              // This value is overwritten when updating the question.
-              .setVersion(VERSION_PLACEHOLDER)
-              .setLifecycleStage(LifecycleStage.DRAFT)
-              .build();
+      questionDefinition = getBuilderWithQuestionPath(roService, questionForm).setId(id).build();
```
Comment (bion) [inline universal-application-tool-0.0.1/app/controllers/admin/QuestionController.java:187]: I like how this controller got simpler!

File: universal-application-tool-0.0.1/app/repository/ApplicantRepository.java [hunk 4]
```diff
@@ -45,23 +52,29 @@ public ApplicantRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext exe
...
+                  ebeanServer
+                      .find(Program.class)
+                      .alias("p")
+                      .where()
+                      .exists(
+                          ebeanServer
+                              .find(Application.class)
+                              .where()
+                              .eq("applicant.id", applicantId)
+                              .eq("lifecycle_stage", LifecycleStage.DRAFT)
```
Comment (bion) [inline universal-application-tool-0.0.1/app/repository/ApplicantRepository.java:66]: Question (non-blocking): looks like applications use the same lifecycle stages but they aren't associated with versions?

File: universal-application-tool-0.0.1/app/repository/ProgramRepository.java [hunk 5]
```diff
@@ -46,6 +46,7 @@ public ProgramRepository(
   public Program insertProgramSync(Program program) {
     program.id = null;
     ebeanServer.insert(program);
+    program.refresh();
```
Comment (bion) [inline universal-application-tool-0.0.1/app/repository/ProgramRepository.java:49]: question (non-blocking): will the in-memory program not have an ID without this line?

File: universal-application-tool-0.0.1/app/repository/ProgramRepository.java [hunk 6]
```diff
@@ -54,30 +55,25 @@ public Program updateProgramSync(Program program) {
...
       this.updateProgramSync(updatedDraft);
       return updatedDraft;
     } else {
-      Program newDraft =
-          existingProgram.getProgramDefinition().toBuilder()
-              .setLifecycleStage(LifecycleStage.DRAFT)
-              .build()
-              .toProgram();
-      insertProgramSync(newDraft);
+      Program newDraft = existingProgram.getProgramDefinition().toBuilder().build().toProgram();
```
Comment (bion) [inline universal-application-tool-0.0.1/app/repository/ProgramRepository.java:71]: question (blocking): why `.toBuilder().build()`?

File: universal-application-tool-0.0.1/app/repository/VersionRepository.java [hunk 7]
```diff
@@ -213,126 +178,40 @@ private BlockDefinition updateQuestionVersions(BlockDefinition block, Transactio
...
-    Optional<Question> existingMaybe =
-        findActiveVersionOfQuestion(questionId, published.getQuestionDefinition().getName());
-    if (existingMaybe.isPresent()) {
-      Question existing = existingMaybe.get();
-      existing.setLifecycleStage(LifecycleStage.OBSOLETE);
-      existing.save();
-    }
-    published.setLifecycleStage(LifecycleStage.ACTIVE);
-    published.save();
+  public void updateProgramsForNewDraftQuestion(long oldId) {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/repository/VersionRepository.java:181]: Suggestion (blocking): please document this method

File: universal-application-tool-0.0.1/app/repository/VersionRepository.java [hunk 8]
```diff
@@ -213,126 +178,40 @@ private BlockDefinition updateQuestionVersions(BlockDefinition block, Transactio
...
-                  return programBean;
-                });
-
-    Streams.concat(draftStream, newDraftsStream)
-        .forEach(
-            program -> {
-              // Create a draft version of each of those programs.
-              updateQuestionVersions(program, transaction);
-            });
+  public void setLive(long versionId) {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/repository/VersionRepository.java:200]: Suggestion (blocking): please document this method

File: universal-application-tool-0.0.1/app/services/program/ActiveAndDraftPrograms.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+import akka.japi.Pair;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import java.util.Optional;
+import models.Program;
+import models.Version;
+
+public class ActiveAndDraftPrograms {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/program/ActiveAndDraftPrograms.java:11]: Suggestion (blocking): please document this class

**Review by carolinedanzi — COMMENTED** [review]
Reviewing slowly, so I'll add comments as I go! Looking good so far!

File: universal-application-tool-0.0.1/app/controllers/admin/AdminVersionController.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  @Secure(authorizers = Authorizers.Labels.UAT_ADMIN)
+  public Result index(Http.Request request) {
+    return ok(versionListView.render(versionRepository.listAllVersions(), request));
+  }
+
+  @Secure(authorizers = Authorizers.Labels.UAT_ADMIN)
+  public Result setVersionLive(long versionId, Http.Request request) {
+    versionRepository.setLive(versionId);
```
Comment (bion) [inline universal-application-tool-0.0.1/app/controllers/admin/AdminVersionController.java:30]: Question (blocking): is there any validation that should take place before calling this? are there any errors that should be handled here? seems unlikely that a bad version ID could come in given it's an admin-only endpoint.

File: universal-application-tool-0.0.1/app/controllers/admin/AdminVersionController.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    return redirect(routes.AdminVersionController.index());
+  }
+}
```
Comment (bion) [inline universal-application-tool-0.0.1/app/controllers/admin/AdminVersionController.java:33]: This is a great example of a controller staying tightly scoped in its responsibilities!

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java [hunk 12]
```diff
@@ -179,8 +175,7 @@ public boolean hasQuestion(long questionId) {
               blockDefinitions().stream()
                   .map(BlockDefinition::programQuestionDefinitions)
                   .flatMap(ImmutableList::stream)
-                  .map(ProgramQuestionDefinition::getQuestionDefinition)
-                  .map(QuestionDefinition::getId)
+                  .map(ProgramQuestionDefinition::id)
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:178]: Question (blocking): are `ProgramQuestionDefinition`'s guaranteed to have the same ID as the `QuestionDefinition`'s they contain?

File: universal-application-tool-0.0.1/app/services/program/ProgramService.java [hunk 13]
```diff
@@ -42,6 +42,8 @@
    */
   ProgramDefinition getProgramDefinition(long id) throws ProgramNotFoundException;
 
+  ActiveAndDraftPrograms getActiveAndDraftPrograms();
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/program/ProgramService.java:45]: Suggestion (blocking): please document this method

File: universal-application-tool-0.0.1/app/services/question/ActiveAndDraftQuestions.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import akka.japi.Pair;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import java.util.Optional;
+import models.Question;
+import models.Version;
+import services.question.types.QuestionDefinition;
+
+public class ActiveAndDraftQuestions {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/ActiveAndDraftQuestions.java:12]: Suggestion (blocking): please document this class

File: universal-application-tool-0.0.1/app/services/question/QuestionServiceImpl.java [hunk 15]
```diff
@@ -21,10 +22,14 @@
 public final class QuestionServiceImpl implements QuestionService {
 
   private QuestionRepository questionRepository;
+  private Provider<VersionRepository> versionRepositoryProvider;
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/QuestionServiceImpl.java:25]: Suggestion (non-blocking): make this and both `private final`

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionService.java [hunk 16]
```diff
@@ -33,6 +33,8 @@
   /** Returns all scalars for this version. */
   ImmutableMap<Path, ScalarType> getAllScalars();
 
+  ActiveAndDraftQuestions getActiveAndDraftQuestions();
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionService.java:36]: Suggestion (blocking): please document this method

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java [hunk 17]
```diff
@@ -17,49 +22,62 @@
...
   private final ImmutableMap<Long, QuestionDefinition> questionsById;
   private final ImmutableMap<Path, QuestionDefinition> questionsByPath;
   private final ImmutableSet<QuestionDefinition> upToDateQuestions;
+  private final ActiveAndDraftQuestions activeAndDraftQuestions;
 
   private Locale preferredLocale = Locale.US;
 
-  public ReadOnlyQuestionServiceImpl(ImmutableList<QuestionDefinition> questions) {
-    checkNotNull(questions);
+  public ReadOnlyQuestionServiceImpl(Version active, Version draft) {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java:36]: Suggestion (non-blocking): append `Version` to `active` and `draft` var names

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java [hunk 18]
```diff
@@ -17,49 +22,62 @@
...
+    checkNotNull(active);
+    checkState(active.getLifecycleStage().equals(LifecycleStage.ACTIVE));
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java:38]: Suggestion (blocking): provide error message in second arg to `checkState`

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java [hunk 19]
```diff
@@ -17,49 +22,62 @@
...
+    checkNotNull(draft);
+    checkState(draft.getLifecycleStage().equals(LifecycleStage.DRAFT));
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java:40]: Suggestion (blocking): provide error message in second arg to `checkState`

File: universal-application-tool-0.0.1/conf/evolutions/default/17.sql [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+create table versions_programs (
+  programs_id bigint,
+  versions_id bigint,
+  primary key (programs_id, versions_id)
+);
+
+insert into versions (lifecycle_stage, submit_time) values ('active', current_timestamp);
+
+insert into versions_questions (questions_id, versions_id)
+  select id, 1 from questions
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/conf/evolutions/default/17.sql:25]: Question (blocking): do we need to migrate applications that have the old 'versions'?

File: universal-application-tool-0.0.1/test/support/TestQuestionBank.java [hunk 21]
```diff
@@ -266,6 +248,17 @@ private Question applicantFavoriteColor(QuestionEnum ignore) {
   private Question maybeSave(QuestionDefinition questionDefinition) {
     Question question = new Question(questionDefinition);
     if (canSave) {
+      // This odd way of finding the active version is because this class is
+      // entirely static and doesn't have room for injection of the Version repository,
+      // because it needs to work without the database.
```
Comment (bion) [inline universal-application-tool-0.0.1/test/support/TestQuestionBank.java:253]: Thought (non-blocking): I think someone is either in-process or already has changed this class to use instance state

Comment (bion) [reply universal-application-tool-0.0.1/test/support/TestQuestionBank.java:253]: fun fact, I drafted a ToTT about avoiding static context in Java and it was mostly well-received by the review committee but one person vetoed it 😞

File: universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+            renderVersionCard(activeVersion),
+            renderHeader("Older Versions"),
+            renderPastVersionTable(olderVersions, request));
+
+    return layout.render(bodyContent);
+  }
+
+  private String renderDateTime(Instant time) {
+    LocalDateTime datetime = LocalDateTime.ofInstant(time, ZoneId.of("America/Los_Angeles"));
+    return datetime.format(DateTimeFormatter.ofPattern("yyyy/MM/dd 'at' h:mm a"));
```
Comment (bion) [inline universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java:73]: Suggestion (blocking): extract `DateTimeFormatter.ofPattern("yyyy/MM/dd 'at' h:mm a")` to a shared constant on `BaseHtmlView`

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java [hunk 23]
```diff
@@ -33,27 +30,28 @@ public ProgramIndexView(AdminLayout layout) {
...
     return layout.render(head(layout.tailwindStyles()), body(contentDiv));
   }
 
-  private Tag maybeRenderPublishButton(
-      ImmutableList<ProgramDefinition> programs, Http.Request request) {
+  private Tag maybeRenderPublishButton(ActiveAndDraftPrograms programs, Http.Request request) {
     // We should only render the publish button if there is at least one draft.
-    if (programs.stream()
-        .anyMatch(program -> program.lifecycleStage().equals(LifecycleStage.DRAFT))) {
+    if (programs.anyDraft()) {
```
Comment (bion) [inline universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java:54]: Question (blocking): if an admin creates a new draft and updates a single question, they should be able to publish that change. Will the programs that depend on that question automatically have drafts created?

File: universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java (hunk 22 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java:73]: Er, actually please exact this method instead, sorry.

**Review by bion — APPROVED** [review]

File: universal-application-tool-0.0.1/app/repository/VersionRepository.java [hunk 24]
```diff
@@ -40,143 +39,89 @@ public VersionRepository(EbeanConfig ebeanConfig, ProgramRepository programRepos
...
+      Version newDraftVersion = new Version(LifecycleStage.DRAFT);
+      ebeanServer.insert(newDraftVersion);
+      return newDraftVersion;
     }
-
-    return 1L;
   }
 
-  private Optional<Question> getLatestVersionOfQuestion(Transaction transaction, long questionId) {
+  public Version getActiveVersion() {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/repository/VersionRepository.java:104]: nit: please add docstring here too

File: universal-application-tool-0.0.1/app/repository/VersionRepository.java [hunk 25]
```diff
@@ -213,126 +178,40 @@ private BlockDefinition updateQuestionVersions(BlockDefinition block, Transactio
...
-    Streams.concat(draftStream, newDraftsStream)
-        .forEach(
-            program -> {
-              // Create a draft version of each of those programs.
-              updateQuestionVersions(program, transaction);
-            });
+  public void setLive(long versionId) {
+    try {
+      ebeanServer.beginTransaction();
+      Version draftVersion = getDraftVersion();
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/repository/VersionRepository.java:203]: question: when will this method be used? Will we end up generating unnecessary draft versions, since if one does not already exist we create a new draft, only to then mark as DELETED?

File: universal-application-tool-0.0.1/app/controllers/admin/AdminVersionController.java (hunk 10 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/controllers/admin/AdminVersionController.java:30]: Any validation that the UX folks think there should be should be in a modal dialog before we hit this point, I think.

File: universal-application-tool-0.0.1/app/models/Version.java (hunk 2 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/models/Version.java:67]: Streams don't have findOne() - this is equivalent to that.

File: universal-application-tool-0.0.1/app/models/Version.java (hunk 1 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/models/Version.java:21]: Done.

File: universal-application-tool-0.0.1/app/repository/ApplicantRepository.java (hunk 4 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/repository/ApplicantRepository.java:66]: Correct, yeah, we already have it so we might as well, I think.

File: universal-application-tool-0.0.1/app/repository/ProgramRepository.java (hunk 5 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/repository/ProgramRepository.java:49]: It gets the ID, but it doesn't get the version table update.  Something odd in ebeans, I suppose!

File: universal-application-tool-0.0.1/app/repository/ProgramRepository.java (hunk 6 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/repository/ProgramRepository.java:73]: Added comment - to clear version information.

File: universal-application-tool-0.0.1/app/services/program/ActiveAndDraftPrograms.java (hunk 9 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/program/ActiveAndDraftPrograms.java:17]: Done.

File: universal-application-tool-0.0.1/app/services/program/ProgramService.java (hunk 13 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/program/ProgramService.java:46]: Done.

File: universal-application-tool-0.0.1/app/services/question/ActiveAndDraftQuestions.java (hunk 14 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/question/ActiveAndDraftQuestions.java:18]: Done - copied from Programs.

File: universal-application-tool-0.0.1/app/services/question/QuestionServiceImpl.java (hunk 15 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/question/QuestionServiceImpl.java]: Done.

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionService.java (hunk 16 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionService.java:36]: Done.

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java (hunk 18 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java]: Done.

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java (hunk 17 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java]: Done.

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java (hunk 19 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionServiceImpl.java]: Done.

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java (hunk 23 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java:57]: Yes - and we test that in the program lifecycle test.

File: universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java (hunk 22 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java]: Done.

File: universal-application-tool-0.0.1/conf/evolutions/default/17.sql [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+
+# --- !Ups
+create table versions (
+  id bigserial primary key,
+  lifecycle_stage varchar not null,
+  submit_time timestamp
+);
+
+create table versions_questions (
+  questions_id bigint,
```
[the remark this answers is not in the document]
Comment (ndmckinley) [reply universal-application-tool-0.0.1/conf/evolutions/default/17.sql]: Since they're in the primary key, we'd be limited to only one null element, but it's not really too important since these are managed by ebeans ManyToMany annotations, so nulls aren't going to get inserted.  I'll add it since it doesn't matter much either way.

File: universal-application-tool-0.0.1/conf/evolutions/default/17.sql (hunk 20 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/conf/evolutions/default/17.sql:25]: Yep, that's what this does - probably!  Hard to test!  But that's what staging is for, haha.  Oh, unless you mean migrate the ability to go back in time ... in that case, I argue "no it is not worth it."

File: universal-application-tool-0.0.1/test/support/TestQuestionBank.java (hunk 21 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/test/support/TestQuestionBank.java]: Yeah - I could refactor it again to inject the version repository, but I'm just straight up not gonna do it, ha.  I will update the comment so it's true now, though!

File: universal-application-tool-0.0.1/conf/evolutions/default/17.sql (hunk 20 above)
Comment (hmhsu) [reply universal-application-tool-0.0.1/conf/evolutions/default/17.sql:25]: Oh I was worrying about the existing applications, but now I see they don't have a version with it, just lifecycle so I guess they won't be directly affected by this change?

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java (hunk 12 above)
Comment (hmhsu) [reply universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:209]: I think so, ProgramQuestionDefinition basically is a wrapper for serializing QuestionDefinition id as json.

File: universal-application-tool-0.0.1/app/models/Version.java (hunk 2 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/models/Version.java:67]: ah bummer

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 27]
```diff
@@ -28,15 +30,18 @@
   private final ProgramRepository programRepository;
   private final QuestionService questionService;
   private final HttpExecutionContext httpExecutionContext;
+  private final Provider<VersionRepository> versionRepositoryProvider;
 
   @Inject
   public ProgramServiceImpl(
       ProgramRepository programRepository,
       QuestionService questionService,
+      Provider<VersionRepository> versionRepositoryProvider,
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:39]: Question (non-blocking): why this can't be just versionRepository?

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionService.java [hunk 28]
```diff
@@ -27,6 +29,12 @@
   /** Returns all repeater question definitions. */
   ImmutableList<RepeaterQuestionDefinition> getUpToDateRepeaterQuestions();
 
+  /** Returns all scalars for this version. */
+  ImmutableMap<Path, ScalarType> getAllScalars();
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionService.java:33]: Nit (blocking): this might be from a merge conflict. It was deleted in a recent PR.

File: universal-application-tool-0.0.1/test/support/TestQuestionBank.java [hunk 29]
```diff
@@ -41,11 +42,10 @@
  * grouping those methods with the same type together.
  */
 public class TestQuestionBank {
-  private static final long VERSION = 1L;
+  private static final Map<QuestionEnum, Question> questionCache = new ConcurrentHashMap<>();
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/test/support/TestQuestionBank.java:45]: Suggestion (blocking): I think this may be caused by merge conflict, we have moved this to be non static to resolve a bug where db is not used.

File: universal-application-tool-0.0.1/test/support/ProgramBuilder.java [hunk 30]
```diff
@@ -26,27 +28,64 @@ private ProgramBuilder(ProgramDefinition.Builder builder) {
...
 
-  /** Creates a {@link ProgramBuilder} with a new {@link Program}. */
-  public static ProgramBuilder newProgram(String name, String description) {
+  /** Creates a {@link ProgramBuilder} with a new {@link Program} in draft state. */
+  public static ProgramBuilder newDraftProgram(String name, String description) {
+    VersionRepository versionRepository = injector.instanceOf(VersionRepository.class);
     Program program = new Program(name, description, name, description);
-    program.setVersion(1L);
+    program.addVersion(versionRepository.getDraftVersion());
+    maybeSave(program);
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/test/support/ProgramBuilder.java:56]: Thoughts (non-blocking): this is not directly related to this change, but maybeSave is really just `save`, it wouldn't work if there is no db.

File: universal-application-tool-0.0.1/test/controllers/applicant/ApplicantProgramsControllerTest.java [hunk 31]
```diff
@@ -56,22 +55,21 @@ public void index_withNoPrograms_returnsEmptyResult() {
...
+    resourceCreator().insertActiveProgram("two");
+    resourceCreator().insertDraftProgram("three");
 
     Result result =
         controller.index(fakeRequest().build(), currentApplicant.id).toCompletableFuture().join();
 
     assertThat(result.status()).isEqualTo(OK);
     assertThat(contentAsString(result)).contains("one");
     assertThat(contentAsString(result)).contains("two");
-    assertThat(contentAsString(result)).doesNotContain("three");
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/test/controllers/applicant/ApplicantProgramsControllerTest.java:69]: Question: does the behavior change here?

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java [hunk 32]
```diff
@@ -43,18 +40,21 @@ public Content render(ImmutableList<ProgramDefinition> programs, Http.Request re
...
-                    programs.stream()
-                        .collect(new GroupByKeyCollector<>(ProgramDefinition::adminName)),
-                    e -> this.renderProgramListItem(e.getKey(), e.getValue(), request)));
+                    programs.getProgramNames(),
+                    name ->
+                        this.renderProgramListItem(
+                            programs.getActiveProgramDefinition(name),
+                            programs.getDraftProgramDefinition(name),
+                            request)),
+                renderNewProgramButton(),
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java:49]: Thought (blocking): this may also be from a merge conflict. These 2 lines have been moved above (line 41).

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java [hunk 33]
```diff
@@ -75,33 +75,24 @@ private Tag renderNewProgramButton() {
...
-    if (activeProgram.isPresent()) {
-      return activeProgram.get();
-    }
-    return programs.stream().findAny().get();
+    return activeProgram.get();
   }
 
   public Tag renderProgramListItem(
-      String name, ImmutableList<ProgramDefinition> programs, Http.Request request) {
-    // TODO: Move Strings out of here for i18n.
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java:95]: Suggestions (non-blocking): would you mind keeping this comment?

File: universal-application-tool-0.0.1/app/views/admin/questions/QuestionsListView.java [hunk 34]
```diff
@@ -111,22 +106,24 @@ private Tag renderAddQuestionLink() {
     return linkButton.with(dropdown);
   }
 
-  private Tag renderSummary(ImmutableList<QuestionDefinition> questions) {
-    return div("Total Questions: " + questions.size())
+  private Tag renderSummary(ActiveAndDraftQuestions activeAndDraftQuestions) {
+    return div("Total Questions: " + activeAndDraftQuestions.getActiveSize())
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/views/admin/questions/QuestionsListView.java:110]: Thought (non-blocking): should this also include draft questions?

File: universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java [hunk 35]
```diff
@@ -0,0 +1,166 @@
...
+  private Tag renderOlderVersionRow(Version olderVersion, Http.Request request) {
+    return tr().withClasses(Styles.BORDER_B, Styles.BG_GRAY_200, Styles.TEXT_LEFT)
+        .with(
+            td(olderVersion.id.toString()),
+            td(renderDateTime(olderVersion.getSubmitTime())),
+            td(String.valueOf(olderVersion.getPrograms().size())),
+            td(String.valueOf(olderVersion.getQuestions().size())),
+            td(
+                new LinkElement()
+                    .setId("view-question-link-" + olderVersion.id)
```
Comment (hmhsu) [inline universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java:98]: Suggestion (blocking): please give a more appropriate id.

**Review by hmhsu — APPROVED** [review]
AWESOME WORK! Looks good to me with minor comments, mostly from merge conflict I think.

File: universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionService.java (hunk 28 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/question/ReadOnlyQuestionService.java:33]: Ah thanks - I saw changes that use it but I see now that I ... need to remove them?  In progress...

File: universal-application-tool-0.0.1/test/support/TestQuestionBank.java (hunk 29 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/test/support/TestQuestionBank.java]: Thanks, fixed!

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java (hunk 12 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:209]: Yep! The code I'm replacing here assumes you got this object from the `Service` which fills in those definitions.

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java (hunk 27 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java]: Oh, probably could!

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java (hunk 32 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java]: Done!

File: universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java (hunk 33 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/views/admin/programs/ProgramIndexView.java:95]: Oh, these are admin strings and we made the decision not to internationalize them, IIRC.

File: universal-application-tool-0.0.1/app/views/admin/questions/QuestionsListView.java (hunk 34 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/views/admin/questions/QuestionsListView.java]: Sure, added!

File: universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java (hunk 35 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/app/views/admin/versions/VersionListView.java]: Whoops!  Thanks!

File: universal-application-tool-0.0.1/test/controllers/applicant/ApplicantProgramsControllerTest.java (hunk 31 above)
Comment (ndmckinley) [reply universal-application-tool-0.0.1/test/controllers/applicant/ApplicantProgramsControllerTest.java]: Hm, no, that should definitely still be there!  Re-adding.
